### PR TITLE
Refactoring special function implementations

### PR DESCRIPTION
**Pull request by hatemhelal**
This patch adds additional implementations of the special functions used in the integral evaluation methods.  Also revisited the `binom_factor` implementation to avoid the `fori_loop`.

The motivation to provide different implementations is to allow comparing across different metrics such as:

* jit tracing time
* run time
* compile time for ipu backend
* ipu execution

A separate PR to come will introduce measurements across these metrics.

### CODE REVIEW

File: test/test_numerics.py [hunk 1]
```diff
@@ -1,14 +1,46 @@
...
+    assert_allclose(factorial2_fori(x), expect)
+    assert_allclose(factorial2_fori(0), 1)
+
+    assert_allclose(factorial2_lookup(x), expect)
+    assert_allclose(factorial2_lookup(0), 1)
+
+
+@pytest.mark.parametrize("binom_func", [binom_beta, binom_fori, binom_lookup])
+def test_binom(binom_func):
+    x = jnp.array([4, 4, 4, 4])
```
Comment (awf) [inline test/test_numerics.py:38]: Maybe test some edge cases too (1,1) (3,3) (3,0) (0,-1) etc

**Review by awf — APPROVED** [review]
Looks good

File: test/test_numerics.py (hunk 1 above)
Comment (hatemhelal) [reply test/test_numerics.py:38]: good idea, I think $3 \choose 3$ and $3 \choose 0$ should already be covered but I've added some more cases involving ones, zeros, and minus ones.
